docs/en/quality_en.md

```diff
@@ -0,0 +1,147 @@
+
+  <img src="../images/quality-control-2.png" alt="quality-control-2.png" width="850" />
+
+## Theories
+
+### Unit Test
+
+Unit test for testing the correctness of a module, a function, or a class, it will guarantee the correctness of module's behavior in a great extend, and serves as the most basic test in the procedure of software development. **Code coverage** is used to measure the completeness of the unit test.  Figure 4 below shows a statistical chart of Curve code coverage, calculated by the tool "lcov (for c++)", and rendered by "genhtml". Other languages such as go also have their own  Coverage statistics tool.  The unit tests of curve modules are under folder "curve/test".
+
+<p align="center">
+    <img src="../images/quality-cov.png" alt="quality-cov.png" width="950" /><br>
+    <font size=3>Figure 4 Curve code coverage result</font>
+</p>
+
+### Integration test
+
+#### necessity
+
+The issues to be considered in the integration test are mainly the problems after the modules are connected:
+
+- Whether the data passing through the module interface will be lost;
+- Whether the combination of sub-functions can meet the expectation;
+- Whether the sub-modules will affect each other;
+- Whether the error accumulation of a single module will be amplified to an unacceptable level; 
+
+Therefore, it is necessary to conduct a integration test to find and eliminate the problems above that may occur in the module connection after the unit test, and finally construct the required software subsystem or system.
+
+#### Content
+
+- **Function test**
+  
+  Function test is to conduct a comprehensive test on the functions provided by the module from user's prospective. Before the test, it is necessary to design sufficient test cases, considering various system states and parameter input to decide whether the module can still work as usual, and return the expected results. In this process, a scientific test cases design method is required to consider various input scenarios, and complete the test process with least cases and execution steps.
+  
+- **Exception test**
+  
+  Different from function test and performance test, exception test is another test for discovering performance bottlenecks caused by exceptions such as system exception, dependencies exception and the exception of the application itself in order to improve the stability of the system. Common exceptions include disk errors, network errors, data errors or program restarts, etc.  We use software to simulate exceptions in the development stage, such as "libfiu" in C++ and "gofail'' in go, which can simulate various exceptions.
+
+- **Scale test**
+  
+  This is to test whether the module can work normally under a certain scale, and whether it will crash or shows exceptions by observing the utilization of system resources. For example, test whether there will be errors when opening a large number of files and whether the memory usage will be too large.
```

Comment:
it will crash or shows exceptions  
-> it will crash or work unexpectedly
test whether there will be errors
-> test whether errors will occurred

Reply:
> it will crash or shows exceptions
> -> it will crash or work unexpectedly
> test whether there will be errors
> -> test whether errors will occurred

done